### PR TITLE
Various DNN improvements (rebase PR61, PR100, PR103, and PR114)

### PR DESCRIPTION
**Pull request by ariostas**
This is a rebase of PRs #61, #100, #103, and #114. They were all DNN-related, so I grouped them together.

One thing I should mention is that I replaced `#pragma unroll N` with `CMS_UNROLL_LOOP`. There is a `CMS_UNROLL_LOOP_COUNT(N)` macro, but it only works with explicit values of `N`. However, it seems like `CMS_UNROLL_LOOP` works equally well as `#pragma unroll N` and we avoid the warnings.

### CODE REVIEW

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h [hunk 1]
```diff
@@ -1,165 +1,154 @@
 #ifndef RecoTracker_LSTCore_src_alpaka_NeuralNetwork_h
 #define RecoTracker_LSTCore_src_alpaka_NeuralNetwork_h
 
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+
 #include "RecoTracker/LSTCore/interface/alpaka/Common.h"
 #include "RecoTracker/LSTCore/interface/ModulesSoA.h"
 #include "RecoTracker/LSTCore/interface/HitsSoA.h"
 #include "RecoTracker/LSTCore/interface/MiniDoubletsSoA.h"
 #include "RecoTracker/LSTCore/interface/SegmentsSoA.h"
```
[the remark this answers is not in the document]
Comment (ariostas) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:10]: Oh you're right, for some reason I thought I couldn't remove them, but I just got confused

**Comment by ariostas** [comment]
Here is the timing comprison on cgpu-1:

```
Before (2e0a3c1)
Total Timing Summary
Average time for map loading = 553.637 ms
Average time for input loading = 7789.05 ms
Average time for lst::Event creation = 0.00217983 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     12.6      0.5      0.3      1.3      1.5      0.4      0.6      0.5      1.3      0.0      19.0       6.0+/-  1.2      21.2   explicit[s=1]
   avg      2.5      0.6      0.6      1.7      1.7      0.5      0.9      0.6      1.8      0.0      11.0       8.0+/-  1.6       6.7   explicit[s=2]
   avg      3.5      1.1      0.9      2.6      2.5      0.6      1.7      1.0      3.1      0.0      17.2      13.1+/-  2.3       4.9   explicit[s=4]
   avg      5.8      1.6      1.2      3.8      3.8      0.7      2.7      1.5      4.5      0.0      25.7      19.1+/-  3.5       4.8   explicit[s=6]
   avg      8.7      2.3      1.5      5.1      5.0      0.9      3.6      2.2      5.8      0.0      35.2      25.6+/-  4.1       4.8   explicit[s=8]

After (1018ea4)
Total Timing Summary
Average time for map loading = 550.404 ms
Average time for input loading = 7773.55 ms
Average time for lst::Event creation = 0.00333786 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     11.6      0.5      0.3      1.3      1.0      0.4      0.6      0.5      1.3      0.0      17.5       5.5+/-  1.1      19.8   explicit[s=1]
   avg      2.5      0.6      0.5      1.6      1.1      0.5      0.8      0.6      1.7      0.0      10.0       7.0+/-  1.4       6.2   explicit[s=2]
   avg      3.4      1.0      0.8      2.5      1.7      0.6      1.6      0.9      2.9      0.0      15.4      11.4+/-  2.2       4.5   explicit[s=4]
   avg      5.6      1.5      1.1      3.6      2.8      0.7      2.5      1.3      4.1      0.0      23.2      16.9+/-  3.1       4.4   explicit[s=6]
   avg      7.7      1.9      1.5      4.8      3.8      0.9      3.6      2.0      5.4      0.0      31.5      22.9+/-  3.4       4.4   explicit[s=8]
```

**Comment by slava77** [comment]
will this be squash-merged?
It's not obvious we need the history of the notebooks and the network weights

**Comment by ariostas** [comment]
> will this be squash-merged?
> It's not obvious we need the history of the notebooks and the network weights

Yeah, this one makes sense to squash-merge

**Pull request merged**